zebrad/src/commands/start/sync.rs

```diff
@@ -174,22 +176,20 @@ where
                         // It indicates that the remote peer does not have any blocks
                         // following the prospective tip.
                         // TODO(jlusby): reject both main and test net genesis blocks
-                        if hashes[0] == super::GENESIS {
-                            tracing::debug!("skipping response that does not extend the tip");
-                            continue;
+                        match hashes.first() {
+                            Some(&super::GENESIS) => {
+                                tracing::debug!("skipping response that does not extend the tip");
+                                continue;
+                            }
+                            Some(_) | None => {}
                         }
 
                         // ExtendTips Step 4
                         //
                         // Combine the last elements of the remaining responses into
                         // a set, and add this set to the set of prospective tips.
-                        let new_tip = *hashes.last().expect("already checked is_empty");
```

Comment:
It's important for the validation of the response to happen *before* any processing of the response.  Otherwise we are processing potentially-invalid responses, which could allow invalid data to slip through the cracks.  In this case, moving the processing (extracting the tip) from here to the top of the match arm opens an edge case, where if the peer responds with `[genesis]`, we would have previously rejected the message but now that invalid data will pass through to the prospective tip set.

The simplest fix for this would be to restore all of the previous code in the match arm (L164-168 and L177-179) and then to change this line to
```
let new_tip = hashes.pop().expect("already checked is_empty");
```
which gives the behavior (tips are removed and not included in `download_set`) expected by your new step 5 implementation.